### ISSUE TITLE
storage: use IngestExternalFile instead of WriteBatches for applying snapshots

### ISSUE THREAD

**Issue by petermattis** [cockroachdb/cockroach#16954]
Sending a snapshot of a range currently involves generating a series of 256KB `WriteBatches` which are then applied on the recipient. Per the recent work to use the RocksDB `IngestExternalFile` functionality for restore, it would likely be significantly faster on the recipient to send an sstable instead. Using `IngestExternalFile` for applying snapshots should be straightforward in comparison to the usage within restore as we don't have any of the complexity of Raft involved. The sender can create the sstable locally and then stream it to the recipient. And because we send a header before streaming the snapshot, we can probe whether the recipient supports receiving an sstable snapshot which will make the migration story easier.

There is no particular urgency here as applying snapshots is not currently a bottleneck. 

@danhhz Do you have benchmark numbers readily available showing the performance difference between `IngestExternalFile` and applying the equivalent `WriteBatches`?

**Comment by tschottdorf** [comment]
One "problem" is that we always put a range tombstone down right before the ingestion, which would force the SSTable into higher layers. That said, it's probably still faster than what we have today.

**Comment by bdarnell** [comment]
Since the receiving side could stream the sstable straight to disk, we could also avoid holding the entire snapshot in memory at once on the recipient. This would be an important step towards being able to increase the default range size from 64MB.

**Comment by tschottdorf** [comment]
I looked into this in light of a privately reported user issue (the fundamental problem in which is that we at some point added a check that prevents large snapshots from being sent).

I was initially hopeful that this was fully straightforward, but now it seems that I've written an RFC draft and think this needs some care. This whole thing seems workable though and worth tackling, though it definitely merits some more preliminary discussion here and then an RFC.

Ok, here we go:

The receiving side will always end up with an SSTable on disk which can then be ingested. On the sender, there are two options:

1. construct an SSTable on and stream that over. Pro: less bandwidth used because SSTable blocks can be compressed.
2. stream the keys in some other format and write them into an SSTable at the receiver. Pro: a little more aligned with the current code, and thus a smaller change.

I'll assume 1. because it seems like the right thing to do™. We have to do some experimentation to decide whether we need to preserve both paths, but I hope not. We can always stream an SSTable, but perhaps we a) avoid flushing it to disk unless it grows large enough and b) avoid ingesting it if it's too small, reading it into a WriteBatch instead. At the end of this, we may also consider making the default range size closer to our target L6 SSTable size (128mb?).

We need to extend `RocksDBSstFileWriter` so that it allows us to access the SSTable as it's being written (right now, it holds it all in memory and only gives it back on `Finish()`). Instead, we want an API like this:

```go
// Add adds the supplied key-value pair into the SSTable. The returned
// boolean is true if the caller may retrieve the next chunk of the SSTable
// via `Fetch` if they desire to do so.
func (fw *RocksDBSstFileWriter) Add(kv MVCCKeyValue) (bool, error)

// Fetch retrieves the data that has been flushed into the SSTable
// but not retrieved yet (by an earlier invocation of Fetch). By calling `Fetch`
// every time `Add` returns `true`, the caller can retrieve the constructed SSTable
// piecemeal.
func (fw *RocksDBSstFileWriter) Fetch() ([]byte, error)

// FinishAndFetch finalizes the SSTable and calls `Fetch`. `Close` must still
// be called, but the instance must not be used other than that.
func (fw *RocksDBSstFileWriter) FinishAndFetch() ([]byte, error)

// Close idempotently releases the underlying resources attached.
func (fw *RocksDBSstFileWriter) Close()
```

Assuming we have that, the [sending side][sendsnapshot] uses an `RocksDBSstFileWriter` in its iteration over the user-space keys and streams straight to the connection (keeping the rate limiting intact).

There's a caveat here, which is that we use a [ReplicaDataIterator] but we really only want to put the main chunk of userspace keys into an SSTable (holes are bad). We should augment the iterator suitably so that we can detect when we're jumping a gap. Other chunks may be large, too, so the most straightforward solution might be to just make the code adaptive from the start: always stream an SSTable, but send a marker RPC every time the chunk changes, so that the recipient can decide whether the current chunk should be ingested or loaded via a `WriteBatch`.

Log entries deserve a mention. These will have to keep special treatment as log entries will have inlined sideloading payloads which [applySnapshot] sideloads again, so we don't ever want to ingest these. Log entries are already plenty special, as they're not rate limited on the [sendsnapshot][sender] and could be quite large due to inlined payloads. As a first pass, it seems reasonable to leave them alone.

As a corollary caveat (and I think the largest complexity with this approach), the recipient has an atomicity problem: we'll definitely end up with something to ingest (the ~64mb-\infty of user data) and something to apply via `WriteBatch`. This implies two constraints, namely ordering: we must only make the range visible when it's been written completely (this implies that the `RangeDescriptor` must be written last, etc) but it also requires us to have a way to recover from an ill-timed crash or I/O error during snapshot application.

There might be a better way to address that, but one I can think off right now is to first write (and sync) a marker key to RocksDB that we delete at the end of a snapshot application; that we make sure to remove all partial state potentially written by an earlier snapshot before writing our own data; and clearing that marker (atomically with or after syncing) the final write which puts down the `RangeDescriptor`. I haven't thought this through, but potentially Raft [tombstone]s could be augmented to carry a sort of "dirty bit" (which can be cleared by applying a snapshot successfully or GC'ing the replica). Recovering from an I/O error in particular will be tricky. We have to remove the replica from the store and also make sure it isn't used (in meaningful ways) by anyone who might be holding on to it. It's easier to recover from a crash as we'll see the dirty bit and get to clean it up eagerly.

It seems pretty bad to render a replica unusable after a failed snapshot application, but keep in mind that the "failure" (assuming no bugs) would be an I/O error, and we generally crash on those intentionally (to avoid undefined behavior). Additionally, a snapshot is either preemptive (in which case there isn't really a Replica) or Raft-trigger (in which case the Replica is so far behind that it is effectively not participating much in the Range, though it does get to vote).

[sendsnapshot]: https://github.com/cockroachdb/cockroach/blob/aae042402ccc78d655cec3bee2c358c8c091afb3/pkg/storage/store.go#L3351
[ReplicaDataIterator]: https://github.com/cockroachdb/cockroach/blob/aae042402ccc78d655cec3bee2c358c8c091afb3/pkg/storage/replica_data_iter.go#L35
[applySnapshot]: https://github.com/cockroachdb/cockroach/blob/aae042402ccc78d655cec3bee2c358c8c091afb3/pkg/storage/replica_raftstorage.go#L687
[tombstone]: https://github.com/cockroachdb/cockroach/blob/aae042402ccc78d655cec3bee2c358c8c091afb3/pkg/roachpb/internal_raft.proto#L37

**Comment by tschottdorf** [comment]
@petermattis am I correct in assuming that ingesting a "everything-covering" SSTable is a terrible idea? Do you think there's any chance of teaching RocksDB to ingest a set of SSTables plus WriteBatches atomically?

**Comment by petermattis** [comment]
> am I correct in assuming that ingesting a "everything-covering" SSTable is a terrible idea?

Yeah, I'm pretty sure that would be very bad.

> Do you think there's any chance of teaching RocksDB to ingest a set of SSTables plus WriteBatches atomically?

I don't know. I think the sstable ingestion is atomic due to writing the manifest. But WriteBatches are atomic due to writing to the WAL. I'm not sure if it if would be possible to connect to the two atomic mechanisms.

**Comment by danhhz** [comment]
An "everything-covering" sstable effectively just means that it will be ingested at L0, which is fine if it's small, but probably bad if it's big.

RocksDB has assured us that it will be Hard(tm) to make an atomic sstable ingestion + writebatch application, but ingesting 2 sstables is already atomic. Is there any way we can break up what you're doing into a large sstable that affects only a small keyrange and a small sstable that covers everything? Perhaps that could be made to work, though it seems to close enough to the margin of what RocksDB has imagined that I'd try it before committing to that plan

**Comment by tschottdorf** [comment]
@danhhz using only SSTables might be one way out. I'm not clear how bad a small everything-covering SSTable would be. BTW, are you saying 2 sstables are atomic or any number of SSTables is atomic? I wonder if it'd be better to ingest ~10 small SSTables and one large one if that makes these small ones "gapless".

**Comment by danhhz** [comment]
any number. IngestExternalFile takes a vector of paths and applies them all atomically

**Comment by danhhz** [comment]
i'm reading over the code again and just noticed that the files you ingest must themselves be mutually non-overlapping. dunno if that's relevant but fyi

it does appear to place each file individually in the lowest level where it doesn't overlap anything, so you should indeed be able to atomically have small, overlappy ones end up in L0 and big, non-overlappy ones in lower levels

One thing you do have to consider in the snapshot case is whether the range of data you're ingesting has any recent writes. If so, you could still end up in the bad case

**Comment by tschottdorf** [comment]
@danhhz using only SSTables is a valid idea. That might work! I think the log entries can be streamlined by changing the API: (receiver permitting), send slim entries and send the files separately. The receiver can first populate the sideloaded storage and then do the ingestion.

**Comment by tschottdorf** [comment]
we'd likely want to send multiple SSTables anyway, right? We have a target size of ~128MB I think in L6, so that seems like the size at which we'd want to start a new one. Or is that not necessary/worse than sending a huge one (for the user data, i.e. without holes)?

Does RocksDB ever have to load SSTables into memory (I hope not)? If so, large SSTables would be problematic anyway. I'm not sure if atomically ingesting N smaller SSTables would be any better, though.

**Comment by petermattis** [comment]
RocksDB streams sstables into memory while compacting. And it loads/pins sstable index blocks into memory for normal reads, but these index blocks are a small fraction of the sstable size. The only downside to large sstables that I'm aware of are they make certain compactions slower which ties up resources that could be used for other (higher priority) compactions. The downside to smaller sstables is we can either run out of file descriptors, or thrash the cache of open sstables.

**Comment by tschottdorf** [comment]
@petermattis in light of that, sticking to ~128mb SSTables seems reasonable?

**Comment by petermattis** [comment]
> @petermattis in light of that, sticking to ~128mb SSTables seems reasonable?

Yes. I'm not aware of any compelling reason to adjust that size right now.

**Comment by tschottdorf** [comment]
I have a POC that doesn't quite work yet, (but should be in a decent shape soon enough). The analysis above seems to be accurate, I haven't seen any additional problems crop up yet. Just a heads up.

**Comment by tschottdorf** [comment]
Link to POC. It does not address any of the problems discussed in [1], but it gets far enough to encounter all of them. 

https://github.com/cockroachdb/cockroach/compare/master...tschottdorf:experiments/snapshotsst

I will look into the RocksDB SST writer next to see whether it exposes a good API for building up SSTables incrementally at the sender (streaming them out bit by bit).

**Comment by petermattis** [comment]
I'm 95% certain the sstables are written in an append-only manner. I'm 100% certain that this was true of sstables inside Google, but it is possible something has diverged in RocksDB. If the sstables are generated in an append-only manner, streaming them from the sender to recipient seems relatively straightforward.

**Comment by tschottdorf** [comment]
> I'm 95% certain the sstables are written in an append-only manner. I'm 100% certain that this was true of sstables inside Google, but it is possible something has diverged in RocksDB. If the sstables are generated in an append-only manner, streaming them from the sender to recipient seems relatively straightforward.

That's good to know, though I'm not sure how it relates to my comment. The way SSTables are constructed is certainly compatible with what we're trying to do. What I was looking into is what APIs there are that we can plumb up, because what we have available in Go at the moment always assumes you're willing to hold the whole SSTable in memory. I'm not worried that there's something here we can't do, just poking around to see whether it's already been done.

**Comment by danhhz** [comment]
> what we have available in Go at the moment always assumes you're willing to hold the whole SSTable in memory.

This was not true before 4d8f1f706a58d21637880e1ddcde0775418f96d0. Should be easy to add the flexibility back.

**Comment by tschottdorf** [comment]
Thanks @danhhz! That's better, though in an ideal world we'd be able to stream the SSTable as it's being created without writing it down to disk first. Unfortunately, having [poked around a bit in RocksDB](https://github.com/facebook/rocksdb/blob/b5c99cc908a7e34fa65d588b2706c33000786935/table/sst_file_writer.cc), it seems that that's just not something they expose in any way. Since we're targeting ~128MB per SSTable, it's probably fine to start with the current code (which sucks it all into memory before streaming it out) and resuscitate 4d8f1f7 when we want to start sending more than one snapshot at once.

**Comment by danhhz** [comment]
Yeah, for streaming it out as you go, there's always writing it to disk and tailing it. You could build your own `rocksdb::Env` implementation, which would let you do basically anything, but I suspect we're a long way from that being worth it.

**Comment by tschottdorf** [comment]
As for setting the SST boundaries right, I was thinking to do the following

- whenever the replicated range iterator enters into a new range, and the first possible key isn't there, add a RocksDB deletion at that key (this will become the smallest key in the SSTable)
- similarly, when leaving a key range, put a deletion tombstone at the upper boundary if there isn't a value.

The latter part is actually tricky because the key range is half-open on the right and there isn't a largest key. We also can't just use the boundary because those usually actually exist in the neighboring range. This then suggests after all that it's better for the receiver to generate the SSTables it needs to cover the remaining space because the sender has no idea what needs to be overwritten. So the sender would indicate the key span for which it is next sending SSTs, and the receiver must (in applySnapshot with raftMu held) generate auxiliary SSTs that simply clear out any existing keys in the gaps (up to two for each "segment" of the range iterator).

For example, if the key ranges are `A=[/Local/Range/10, /Local/Range/11)` and `B=[/Global/a, /Global/c)` then the sender may send 12 SSTables for `A` that in total cover `A' =
 [/Local/Range/10/c, /Local/Range/10/q)` and 100 SSTables for `B` that in total cover `B' =
 [/Global/f, Global/g)`. Note that in the example below I assume that the sender doesn't do the deletion on the start key (this would be a strict optimization, but perhaps it's not worth the symmetry breaking).

The receiver would receive `A'` (and write down the 12 corresponding SSTs) and `B'` (again writing down the SSTs). Then, inside of `applySnapshot` with `raftMu` held, it'd iterate over `[/Local/Range/10, /Local/Range/10/c)` and generate an SSTable that clears out any value found, then deal with the span `[/Local/Range/10/q, /Local/Range/11)` similarly, emitting an SSTable, and similarly for `[/Global/a, /Global, f)` and `[/Global/g, /Global/q)`. Adding these additional SSTable files to the ones originally received from the sender, everything can be ingested.

Of course it will be awkward when these gap SSTables actually contain a lot of keys that need to be removed (as would happen, for example, when the snapshot contains large deletions not observed by the replica being caught up), and may need to be chunked (again, to not blow up memory). The batching can be abstracted out so no problem, but it seems a little more complex than one would hope.

Perhaps there's a better way. I *think* (but don't understand this well enough) that range deletion tombstones live somewhere above raw SSTables, so we can't use them here.

I added a second commit to my [experiment branch](https://github.com/tschottdorf/cockroach/commits/experiments/snapshotsst) but I don't think there's anything new in there (and it doesn't work; I stopped to let @nvanbenschoten pick up the work).

**Comment by petermattis** [comment]
> Perhaps there's a better way. I think (but don't understand this well enough) that range deletion tombstones live somewhere above raw SSTables, so we can't use them here.

I'm not sure what you mean by this. Range deletion tombstones live inside sstables in the "range deletion block". Care is taken to propagate the range deletion tombstones during compaction to any new sstables for which there is overlap.

**Comment by tschottdorf** [comment]
What I mean concretely is that if my [reading is correct](https://github.com/facebook/rocksdb/blob/b5c99cc908a7e34fa65d588b2706c33000786935/table/sst_file_writer.cc#L88), then range deletions can't be written into SSTs. Perhaps this isn't the actual code that's used during compactions, and support for range deletions can be added in. I think that would be preferable, to the complexity I outlined above, but I'm not familiar enough with the RocksDB code to gauge how difficult it would be.

**Comment by petermattis** [comment]
You're reading looks correct to me. I wonder why that limitation exists as range deletion tombstones [definitely exist in sstables](https://github.com/facebook/rocksdb/blob/master/table/block_based_table_builder.cc#L431-L440).

**Comment by danhhz** [comment]
> Perhaps this isn't the actual code that's used during compactions

Correct, this is specifically for preparing sstables to be ingested via IngestExternalFile.

Range deletions were added after IngestExternalFile and there was some initially some correctness issues between the two that we got them to fix. Which means I would not be surprised if the range deletion implementor wasn't familiar with the IngestExternalFile stuff.

**Comment by tschottdorf** [comment]
Ok. But that means we should be able to use (an updated version of) `SstFileWriter` just fine to create SSTables that only contain a range deletion each, filling up the gaps (?).

**Comment by danhhz** [comment]
That I can't say for sure. We'd have to ask RocksDB

**Comment by tschottdorf** [comment]
Posted the question: https://www.facebook.com/groups/rocksdb.dev/permalink/1508823992549443/

**Comment by nvanbenschoten** [comment]
I just posted that question to the mailing list as well https://groups.google.com/forum/#!topic/rocksdb/9cWXn8WsMH0.

**Comment by nvanbenschoten** [comment]
I opened up an issue in RocksDB's repo to track `DeleteRange` support in `SstFileWriter`: https://github.com/facebook/rocksdb/issues/3391. I made an attempt to do this myself today, but I think the design is complicated enough that it's going to need some buy in from the RocksDB team.

**Comment by nvanbenschoten** [comment]
With the elimination of the range size limit on splits in https://github.com/cockroachdb/cockroach/pull/21777, this is less immediately pressing. I'm still planning on following https://github.com/facebook/rocksdb/issues/3391 through to completion, at which point we can revisit this early on in the 2.1 timeframe.

**Comment by jeffrey-xiao** [comment]
Closed by #38932.
